### ISSUE TITLE
Save folded directories in project

### ISSUE THREAD

**Issue by eth0net** [zed-industries/zed#19898]
### Check for existing issues

- [X] Completed

### Describe the feature

I'd love to see the option to save and restore the folded state of directories when reloading or switching projects.

In some of my projects I have a few extra folders added to reference related code but don't always need them so when opening a project I have to fold the ones I don't want right now (but want to keep around) each time I reload or switch projects.

I'd also suggest adding a setting to toggle this feature, allowing users control over it.

**Comment by osiewicz** [comment]
Can you be a bit more specific? Would you like us to track which worktrees are folded at the time you close a project, or would you like us to track which folders in a single worktree were unfolded?

**Comment by lode** [comment]
I have the same/similar wish.

> Would you like us to track which worktrees are folded at the time you close a project, or would you like us to track which folders in a single worktree were unfolded?

What do you mean with worktree? Do you mean a root-level folder in the project? (The root-level folder added when using "Add folder to project"?)

I would prefer having the whole folder pane toggle state remembered when closing a project. But only the root-level folders would be good enough for me.

**Comment by eth0net** [comment]
Yeah the root-level folder in the project would be the main thing I'm after with saving, but the entire list of folded/unfolded directories could be nice.

The main issue I have is that all root-level directories are unfolded by default, then child directories are only unfolded for open files in the editor, and I don't want all root-level unfolded each reload :)

**Comment by eth0net** [comment]
> I think they mean something like the workspace concept in VSCode(Image attached) and then the "Save workspace as..." option in the file menu ![image](https://private-user-images.githubusercontent.com/120726854/381575520-57815ee3-32b0-4f89-8ce1-57cf9b4e33f5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA3MjA4MTUsIm5iZiI6MTczMDcyMDUxNSwicGF0aCI6Ii8xMjA3MjY4NTQvMzgxNTc1NTIwLTU3ODE1ZWUzLTMyYjAtNGY4OS04Y2UxLTU3Y2Y5YjRlMzNmNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTA0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEwNFQxMTQxNTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mMWU5ZTA3MTEzYTQ0NzgzZDc0NzQ3YzIxYjIwYzVjNmY2Nzg4N2RjMmZkMDRjNDJiMWY2YjI2NTg1ZjcxODI1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.-VAP7zMgnk3UMueO5-UHG0k_htIeu1WN-xVt6Rzl91c)
> 
> Edit - Here the 2 folders added to the workspace are 2 different git repo's for one of my project

Not quite what I'm after since Zed already saves the list of directories as a project in the database, I don't need an extra file lying around :D 

But iirc VS Code workspaces do save the folded state for every directory.

**Comment by eth0net** [comment]
As an example, here's a multi-directory project I have right now. I currently only have files from the top directory open and as such only have that directory unfolded for now.

![Screenshot 2024-11-04 at 11 48 37](https://github.com/user-attachments/assets/58de25b4-3976-413a-8e48-af8e46369c03)

After a reload, all of the other directories unfold again like this which can be frustrating when quickly restarting for Zed updates before continuing to work.

![Screenshot 2024-11-04 at 11 52 27](https://github.com/user-attachments/assets/6e580cca-a104-42e7-ba98-061429aa7d3b)

**Comment by osiewicz** [comment]
Ok, yeah, worktree is essentially:
> Do you mean a root-level folder in the project? (The root-level folder added when using "Add folder to project"?)

Since we already store the list of worktrees in db, we might as well store their fold state.

**Comment by lode** [comment]
If we're looking for a quick fix, maybe we can change the default to fold root-level folders instead unfold them. And only unfold the ones with open files ? and/or only the first root-level folder if there are no files open.

This is not very good, since for example for one of my projects my most used folder is not the first one. But better than having to fold all root folders before getting some overview.

**Comment by jtmueller** [comment]
@osiewicz While you're at it, could you also save the order of the top-level folders in a project, as well as the fold state? If I'm opening a multi-repo project, I prefer to order my top-level folders by importance or sequence of events or frontend-before-backend, and it gets a bit frustrating that the UI will let you drag the top-level folders to change the order, but then doesn't remember this order.

**Comment by eth0net** [comment]
> [@osiewicz](https://github.com/osiewicz) While you're at it, could you also save the order of the top-level folders in a project, as well as the fold state? If I'm opening a multi-repo project, I prefer to order my top-level folders by importance or sequence of events or frontend-before-backend, and it gets a bit frustrating that the UI will let you drag the top-level folders to change the order, but then doesn't remember this order.

This was previously working from the change I put in to allow reordering worktrees, and I have already submitted a fix for it 🚀 #39934

**Comment by JosephTLyons** [comment]
Hello,

We’re closing this issue as part of a cleanup of older feature requests.

Going forward, new feature requests start as GitHub Discussions and only become Issues once the direction is agreed on. This helps ensure that Issues represent scoped, ready-to-build work and makes it easier for contributors to find work that’s ready to pick up. You can read more in our [Feature Request Guidelines](https://github.com/zed-industries/zed/discussions/47963).

As part of this transition, we’re closing feature request Issues that are older than 6 months and have fewer than 10 upvotes.

If this feature hasn’t been implemented and is still important to you, we encourage you to open a new feature request Discussion [here](https://github.com/zed-industries/zed/discussions/new?category=feature-requests). That’s where we track and evaluate new requests.

Thanks for helping shape the future of Zed.

**Comment by eth0net** [comment]
I implemented this here https://github.com/zed-industries/zed/pull/56521
